crates/pera-framework/packages/pera-system/sources/dwallet_2pc_mpc_ecdsa_k1.move

```diff
@@ -0,0 +1,1045 @@
+    /// - Validates the linkage between dWallet components.
+    /// - Verifies that the number of `hashed_messages`, `message_approvals`, and
+    ///   `centralized_signed_messages` are equal.
+    /// - Emits a `StartSignEvent` with the hashed message, presign outputs,
+    ///   and additional metadata.
+    ///
+    /// ### Emits
+    /// - `StartBatchedSignEvent`:
+    ///   - Contains the session details and the list of hashed messages.
+    /// - `StartSignEvent`:
+    ///   - Includes session details, hashed message, presign outputs,
+    ///     and DKG output.
+    ///
+    /// ### Aborts
+    /// - **`EDwalletMismatch`**: If the `dwallet` object does not match the ID
+    ///   in the `Presign` object.
+    /// - **`EApprovalsAndMessagesLenMismatch`**: If the length of the `hashed_messages`
+    ///   does not match the length of the `message_approvals`.
+    /// - **`ECentralizedSignedMessagesAndMessagesLenMismatch`**: If the length of
+    ///   `hashed_messages` does not match the length of `centralized_signed_messages`.
+    /// - **`EMessageApprovalDWalletMismatch`**: If the DWalletCap ID does not match
+    ///   the expected DWalletCap ID for any of the message approvals.
+    /// - **`EMissingApprovalOrWrongApprovalOrder`**: If the approved messages are not
+    ///   in the same order as the `hashed_messages`.
+    ///
+    /// ### Parameters
+    /// - `dwallet_cap`: The capability associated with the dWallet.
+    /// - `hashed_messages`: The list of hashed messages to be signed.
+    /// - `message_approvals`: The approvals for the messages.
+    /// - `presign`: The presign object containing intermediate outputs.
+    /// - `dwallet`: The dWallet object.
+    /// - `centralized_signed_messages`: The list of centralized signatures.
+    /// - `presign_session_id`: The session ID of the presign process.
+    /// - `ctx`: The mutable transaction context.
+    public fun sign(
+        message_approvals: &mut vector<MessageApproval>,
+        mut hashed_messages: vector<vector<u8>>,
+        mut presigns: vector<Presign>,
+        dwallet: &DWallet<Secp256K1>,
+        mut centralized_signed_messages: vector<vector<u8>>,
+        ctx: &mut TxContext
+    ) {
+        let messages_len: u64 = vector::length(&hashed_messages);
+        let presigns_len: u64 = vector::length(&presigns);
+        let approvals_len: u64 = vector::length(message_approvals);
+        let centralized_signed_len: u64 = vector::length(&centralized_signed_messages);
+        assert!(messages_len == approvals_len, EApprovalsAndMessagesLenMismatch);
+        assert!(messages_len == centralized_signed_len, ECentralizedSignedMessagesAndMessagesLenMismatch);
+        assert!(messages_len == presigns_len, EPresignsAndMessagesLenMismatch);
+        let expected_dwallet_cap_id = get_dwallet_cap_id(dwallet);
+        let batch_session_id = object::id_from_address(tx_context::fresh_object_address(ctx));
+        event::emit(StartBatchedSignEvent {
```

Comment:
Im worried of sending events before security checks happen. You should only do that after all messages are verified to be approved